packages/langchain-graph-retriever/src/langchain_graph_retriever/adapters/astra.py

```diff
@@ -37,61 +37,62 @@
 _MAX_RETRIES = 3
 
 
-class _QueryHelper:
-    def __init__(
-        self,
-        codec: _AstraDBVectorStoreDocumentCodec,
-        user_filter: dict[str, Any] | None = {},
-    ) -> None:
-        self.codec = codec
-        self.encoded_user_filter = (
-            codec.encode_filter(user_filter) if user_filter else {}
-        )
-
-    def encode_filter(self, filter: dict[str, Any]) -> dict[str, Any]:
-        encoded_filter = self.codec.encode_filter(filter)
-        return self._with_user_filter(encoded_filter)
+def _extract_queries(edges: set[Edge]) -> tuple[dict[str, Iterable[Any]], set[str]]:
+    metadata: dict[str, set[Any]] = {}
+    ids: set[str] = set()
 
-    def _with_user_filter(self, encoded_filter: dict[str, Any]) -> dict[str, Any]:
-        if self.encoded_user_filter:
-            return {"$and": [encoded_filter, self.encoded_user_filter]}
+    for edge in edges:
+        if isinstance(edge, MetadataEdge):
+            metadata.setdefault(edge.incoming_field, set()).add(edge.value)
+        elif isinstance(edge, IdEdge):
+            ids.add(edge.id)
         else:
-            return encoded_filter
+            raise ValueError(f"Unsupported edge {edge}")
 
-    def create_ids_query(self, ids: list[str]) -> dict[str, Any]:
-        if len(ids) == 0:
-            raise ValueError("IDs should not be empty")
-        elif len(ids) == 1:
-            # We need to avoid re-encoding in this case, since the codecs
-            # don't recognize `_id` and get confused (creating `metadata._id`).
-            return self._with_user_filter({"_id": ids[0]})
-        elif len(ids) < 100:
-            return self._with_user_filter({"_id": {"$in": ids}})
-        else:
-            raise ValueError("IDs should be less than 100, was {len(ids)}")
+    return (cast(dict[str, Iterable[Any]], metadata), ids)
 
-    def create_metadata_query(
-        self, metadata: dict[str, Iterable[Any]]
-    ) -> dict[str, Any] | None:
-        metadata = {k: v for k, v in metadata.items() if v}
-        if not metadata:
-            return None
 
-        parts = []
-        for k, v in metadata.items():
-            # If there are more than 100 values, we can't create a single `$in` query.
-            # But, we can do it for each batch of 100.
-            for v_batch in batched(v, 100):
-                batch = list(v_batch)
-                if len(batch) == 1:
-                    parts.append({k: batch[0]})
-                else:
-                    parts.append({k: {"$in": batch}})
-
-        if len(parts) == 1:
-            return self.encode_filter(parts[0])
+def _queries(
+    codec: _AstraDBVectorStoreDocumentCodec,
+    user_filters: dict[str, Any] | None,
+    *,
+    metadata: dict[str, Iterable[Any]] = {},
+    ids: Iterable[str] = (),
+) -> Iterator[dict[str, Any]]:
+    encoded_user_filters = codec.encode_filter(user_filters) if user_filters else {}
+
+    if user_filters:
+        user_filters = codec.encode_filter(user_filters)
```

Comment:
definition of `encoded_user_filters` can go into the `if` branch; and in the `if` branch, in turn, `user_filters = ...` seems superfluous